packages/gatsby/src/redux/nodes.ts

```diff
@@ -512,29 +500,20 @@ export const getNodesFromCacheByValue = (
     const values = filterCache.meta.valuesAsc as Array<FilterValue>
     // It shouldn't find the targetValue (but it might) and return the index of
     // the two value between which targetValue sits, or first/last element.
-    const point = binarySearch(values, filterValue)
-    if (!point) {
-      return undefined
-    }
-    const [pivotMin, pivotMax] = point
-    // Each pivot index must have a value and a range
-    // The returned min/max index may include the lower/upper bound, so we still
-    // have to do lte checks for both values.
-    let pivotValue = values[pivotMax]
-    if (pivotValue > filterValue) {
-      pivotValue = values[pivotMin]
-    }
-
-    // Note: the pivot value _shouldnt_ match the filter value because that
-    // means the value was actually found, but those should have been indexed
-    // so should have yielded a result in the .get() above.
-
-    const [exclPivot, inclPivot] = ranges!.get(pivotValue) as [number, number]
+    const idx = binarySearch(values, filterValue)
 
-    // Note: technically, `5 <= "5" === true` but `5` would not be cached.
-    // So we have to consider weak comparison and may have to include the pivot
-    const until = pivotValue <= filterValue ? inclPivot : exclPivot
-    return new Set(nodes!.slice(0, until))
+    if (idx >= 0) {
```

Comment:
I was confused by the further tests on the returned indexes from `binarySearch` and the indirection through `ranges!.get(pivotValue)`. This changed code just returns the set based on the range found based on the binary search. This maybe misses some cache interaction or something? Happy for guidance here...

Reply:
The reason for this indirection is that this approach is to be reused for `<` and `<=`. One needs to slice up to the pivot, the other will need to also include all the nodes for given value (the values array is most likely not unique).

I think that this is necessary for both cases and I'm surprised this would work for you.

For `lt` you need it for the case like `[1,2,2,2,2,3]`, where binary search would just return the middle index, which is not what you should slice on because you want to slice `[1]`, not `[1,2,2]`.
For `lte` you would want to receive `[1,2,2,2,2]`, not  `[1,2,2]`.

The indirection takes the value, looks up the precomputed range for all nodes with that value, and then slice based on that range, and whether the pivot was found (include the range) or not (exclude the range).

I hope this clarifies it a bit. If that makes sense, can you help me to confirm whether or not your approach should work? Because I feel it shouldn't, but I expect that you ran the tests and it did work.

(Looking at it now, and considering a 1:1 index match is kinda useless, the set of ordered values could be a _set_ (of uniques) which would speed up the binary search in some cases)